### ISSUE TITLE
Support for chunked transfer encoding

### ISSUE THREAD

**Issue by carlos-verdes** [bigskysoftware/htmx#1911]
Hi, I have a server that is responding with a http chunked response to load data as it's available (also it reduces the consumption of memory on the server):
https://en.wikipedia.org/wiki/Chunked_transfer_encoding

When I make a direct call with my browser I can see how the results pop-up as they are flushed from the server (as expected) however if I use `hx-get` I don't see anything in the screen until the full response is sent (not the expected behavior)

I see there is support for SSE and Web Sockets, is there a plan to support this feature also?

**Comment by carlos-verdes** [comment]
I use that API to read from the server when doing SPA development, that's why I'm asking if HTMX will support this feature or not

The good thing about it is you don't need to change the protocol and most of HTTP and actually HTMX just works, the only problem is behavior is not as expected (HTMX waits for full response to be sent before rendering anything back to the user)

**Comment by joetifa2003** [comment]
I really need this feature (http streaming).

**Comment by sjc5** [comment]
It would be lovely to see this supported

**Comment by jtoppine** [comment]
+1

Example use case: Streaming search results. Instead of complex "infinite scroll" or other types of convoluted result delivery schemes, we could just start pushing results to the client as soon as we get first hits from the database - client could start loading and rendering images etc for the first entries in listing right away. Ohh, it would be so straightforward and beautiful, so old school in the bestest of ways.

**Comment by carlos-verdes** [comment]
I send all collections from my backend using streams to avoid also memory pressure on big collections so for me is a natural thing to do

**Comment by alexpetros** [comment]
Had to close the above PR but I think the existing extension mechanism should be more than sufficient to implement this. Would love if someone wanted to take that on.

**Comment by douglasduteil** [comment]
@alexpetros I'm on it ;)

**Comment by carlos-verdes** [comment]
Are you doing extension for this @douglasduteil ?
Can you share link to the PR when ready?

**Comment by douglasduteil** [comment]
# 🎉 https://github.com/douglasduteil/htmx.ext...chunked-transfer 🎉 

\to @carlos-verdes it's christmas time again 🎄 

## Install

```sh
$ npm install htmx.ext...chunked-transfer
```

```html
<script src="https://unpkg.com/htmx.ext...chunked-transfer/dist/index.js"></script>
```

## Usage

```html
<body hx-ext="chunked-transfer">
  ...
</body>
```


--- 

:warning: It's a very early version that I'm not using myself

**Comment by mattbrandman** [comment]
I don't know if there is a plan to add this into the base of HTMX or rely on extensions but I thought this is an example of the functionality folks are looking for https://livewire.laravel.com/docs/wire-stream. The ability to append vs replace is a nice addition as well. 

I would bet a good deal of this is around streaming back AI based content. While using SSE or webhooks is an option it adds complexity to infra depending on the surrounding infra. The chunked transfer encoding feels clean because once all the data is sent the connection is closed vs SSE needing to be replaced to be closed with no real "polite close" or having to deal with websockets in general. 

The other simplicity comes from not having to deal with channels for SSE or websockets on the server for multiple client, when you want to send back to only the sender the websocket / sse solutions feel heavyweight.

**Comment by JEBailey** [comment]
I have an ES5 version of an extension that supports the chunked encoding that I've been using internally at the company I work for.

https://github.com/JEBailey/htmx/blob/master/src/ext/chunked.js

**Comment by alexpetros** [comment]
So I originally closed #2101 because 2.0 was coming up and we weren't going to make that happen in time. I'm seeing some compelling use-cases and @douglasduteil's extension looks like it's been working. Are people using it? What's the case for including this in core?

**Comment by fabge** [comment]
one usecase would be a fairly simple chatbot app which supports streaming.
no need for websockets, no need for server-sent events, no need for keeping a connection to the server. 
you could simply make a request to the server, the server sends a `Transfer-Encoding: chunked` response, which will then incrementally be swapped/added into the respective chat bubble.

**Comment by runeksvendsen** [comment]
> So I originally closed #2101 because 2.0 was coming up and we weren't going to make that happen in time. I'm seeing some compelling use-cases and @douglasduteil's extension looks like it's been working. Are people using it? What's the case for including this in core?

Hi @alexpetros,

Yes, I'm using @douglasduteil's extension in https://github.com/runeksvendsen/haskell-function-graph and it's working for me. Thank you @douglasduteil!

The case for including it in core is that it solves a very generic problem: you don't want the user to wait for the very last part of your page to be received until the first part of the page is shown. The larger the time difference between the backend having the first and last result available the worse this problem is.  
In my case, I have a page that includes in the following order: (1) a list of results, where the first result is usually available to the backend very quickly (within a few milliseconds) and the last result can take an additional ~second to become available; followed by (2) an SVG graph that's slow to generate, because it calls out to a CLI executable. Without this extension, the user has to wait around ~two seconds to see the first results, even though they're available to the backend (and sent to the client) almost immediately.

**Comment by jph00** [comment]
@alexpetros a lot of our users are asking for chunked transfer support to make it easier to stream responses from language models, and similar tasks.

**Comment by alexpetros** [comment]
@jph00 have you tried the extension? I would like to get some feedback on how the extension is working for people

**Comment by jph00** [comment]
> Could someone please explain how this would be useful as opposed to using the existing SSE extension? I'm surely missing something fundamental about how they work

I think basically the benefit is that it's simpler.

**Comment by schungx** [comment]
Incidentally I have a couple of pages where the user can (potentially) load a few thousand rows' worth of data on-screen.

How can I setup the extension such that it, sorta, stream like 50 rows at a time?

**Comment by mattbrandman** [comment]
I'm using the extension and it works well for my use case there's a few points that break like trying to use templates with declarative shadowdom doesn't seem to render the initial components https://templ.guide/server-side-rendering/streaming. Though that may have to do with how swapping works read somewhere that it doesn't work with innerhtml call but may be wrong. 

Having something official just means one less extension, potentially more hooks, and less of a chance of breaking with future updates. Am also using it with LLM streaming like others!

**Comment by runeksvendsen** [comment]
> > Could someone please explain how this would be useful as opposed to using the existing SSE extension? I'm surely missing something fundamental about how they work
> 
> I think basically the benefit is that it's simpler.

Also:

* Much better backwards compatibility (only requires HTTP/1.1)
* Doesn't require JavaScript 
* Isn't broken for HTTP/1.1 (see warning at https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events)

**Comment by mattbrandman** [comment]
> > > > Could someone please explain how this would be useful as opposed to using the existing SSE extension? I'm surely missing something fundamental about how they work
> > > 
> > > I think basically the benefit is that it's simpler.
> > 
> > Also:
> > 
> > * Much better backwards compatibility (only requires HTTP/1.1)
> > * Doesn't require JavaScript 
> > * Isn't broken for HTTP/1.1 (see warning at https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events)
> 
> Thanks! Can you do all the same things with chunked transfer encoding as you can with sse? Specifically,
> 
> - keep a long-lived connection open and periodically send messages to the browser? 
> - allow for JavaScript (eg htmx or any other script, service worker etc) to initiate and receive/process the messages? 
> 
> 

I don't believe so but I think part of the niceness is not needing to manage sse connections on the server. Once the connection is closed it's done no need to keep it around.

**Comment by runeksvendsen** [comment]
> Thanks! Can you do all the same things with chunked transfer encoding as you can with sse?

No, you definitely cannot. Chunked transfer encoding is just a way to transfer data from a server to a client (and not the other way around).

That's why it doesn't require JavaScript and supports ancient browsers.

**Comment by fabge** [comment]
> That's why it doesn't require JavaScript and supports ancient browsers.

which is exactly why it should be support by default 🙏

**Comment by fabge** [comment]
sse - just as websockets - establishes long lived connections to the server.

let's take the example of a simple chatbot with a backend deployed on a serverless platform (e.g. aws lambda) where the backend runs just a couple of millseconds at a time (up to a couple of minutes at most).

if we want to have a streaming response, the simplest approach would be to use chunked transfer of the content. this would mean not needing any persistent connection (websockets or sse) between the client and server.

e.g. it would be a simple POST request with the message from the client side and a chunked transfer of the response from the server.

**Comment by fabge** [comment]
yes! and also reducing complexity when the only thing you want is a streaming response. sse and websocket are quite intimidating at first i think.

**Comment by runeksvendsen** [comment]
> But htmx IS JavaScript... And given that some have confirmed that SSE is a more robust and functional protocol than chunked transfers, what's the point of this request?
> 
> [...]
> 
> I'm just very perplexed, but quite open to being educated on what I'm missing about why this is useful, let alone needed.

The main reason I think it should be included in HTMX core is that not including it breaks [Progressive Enhancement](https://htmx.org/docs/#progressive_enhancement) for chunked transfer encoding. 

As I explain in my duplicate issue (https://github.com/bigskysoftware/htmx/issues/2789), the degree of brokenness varies depending on how long it takes the server to close the connection, but in the worst case it's a difference between a _boosted_ link never displaying any content if HTMX is enabled while content is displayed immediately in case it's disabled.

**Comment by wjkoh** [comment]
I found another HTMX extension that supports chunked transfer encoding: https://github.com/alarbada/htmx-stream. It basically enables streaming with replace behavior, not append. Thanks to @alarbada, it's very simple to display progressively generated content over a single GET or POST request.  I would really love it if this HTTP/1.1 functionality were supported by HTMX by default. That way, we could add another simple but powerful layer of interactivity to HTMX web apps without any JavaScript!

**Comment by alarbada** [comment]
Hello @wjkoh. My very much alpha project works somewhat similar to sse, but I did have the difficulties that @carlos-verdes said.

My solution is much easier to maintain, as it is stateless.

**Comment by carlos-verdes** [comment]
@alarbada can you mention which framework are you using now?
I really like htmx, just having issues with this specific point.

**Comment by carlos-verdes** [comment]
@alarbada you saved my life honestly, I copy/pasted your code and just changed behavior to append instead of full rewrite, now I can just send chunked responses and it works as expected!

**Comment by alarbada** [comment]
@carlos-verdes man I changed my stacks back and forth this year, the current thing I've settled with is go + my own thing that compiles handlers into ts sdk + astro with htmx and solidjs.

I know it looks like a pile of random tools but in my pet project works like wonders :D

htmx is my spa router, solidjs is for client side interactivity, astro is the backbone of all, go is there because I hate backend js with a passion.

If I was to program this streaming thing with this stack I would just stream jsons in go and manually parse them in solidjs in this setup, and because htmx doesn't reload the page my solidjs state is still there, _even if the solidjs html is replaced by htmx_.

**Comment by wjkoh** [comment]
Hi @1cg, I just read [The future of htmx](https://htmx.org/essays/future/) and thanks for your insights! I agree with many points, and "No New Features as a Feature" resonates deeply with me. By any chance, do you think it's worth having HTMX support Chunked Transfer Encoding before it goes into the stable state? Websocket and SSE are not always the best choice, and HTTP/2 and its streaming feature seem too far out at the moment. Chunked Transfer Encoding might be the best standard bet we have for ChatGPT-like text streaming. Would you mind sharing your opinion? Best, Woojong

**Comment by wjkoh** [comment]
Just FYI, gRPC-gateway also uses Chunked Transfer Encoding to implement the server-to-client streaming. Check out https://github.com/grpc-ecosystem/grpc-gateway/blob/4aa9147d64085aabcd2af3dc59c02c60cf5e365d/runtime/handler.go#L31 .

**Comment by alexpetros** [comment]
Nothing in the future of htmx essay precludes using chunked transfers (since it would basically be an invisible upgrade). The thing that keeps chunked transfers from happening is that that htmx is that, as others have pointed out, htmx waits for the whole response to come in before performing the swap, which is simpler to code than a buffered stream. 

I don't think we have any ideological objection to it, other than needing it to work in a way that doesn't cause regressions or add major complexity to the code. That having been said... I haven't seen any compelling reasons why the @douglasduteil  extension isn't sufficient. Maybe we just need to promote it alongside some of the other major ones.

**Comment by runeksvendsen** [comment]
@alexpetros the most compelling argument I've seen is that it breaks progressive enhancement, which HTMX claims to support. Ie. *boosting* a link changes the behavior of a web page if the endpoint delivers the response using chunked transfer encoding.

**Comment by alexpetros** [comment]
> boosting a link changes the behavior of a web page if the endpoint delivers the response using chunked transfer encoding

The server is something you control, not a fixed entity. htmx doesn't support chunked transfer encoding, so if you're using it for non-boosted links, then you should turn it off it you detect a boosted one. It doesn't "break" anything, it's just something that htmx doesn't currently support.

This also isn't an argument for why adding `hx-ext` isn't sufficient. It seems like that *does* work fine for everyone.

**Comment by wjkoh** [comment]
Now that htmx 4.0 switches to fetch(), can we get htmx to support chunked transfer encoding as well? https://four.htmx.org/

**Comment by carlos-verdes** [comment]
@wjkoh from the release notes I see this will be supported:

<img width="706" height="265" alt="Image" src="https://github.com/user-attachments/assets/1f67e0a7-8cbf-45e1-9f2f-f64eef0cd72e" />

**Comment by Lanayx** [comment]
@1cg Can chunked encoding support make it to fourth version or become default extension?

**Comment by 1cg** [comment]
Sounds like a great feature for an extension in 4

/// big sky software ///

**Comment by 1cg** [comment]
OK folks, we now, finally, have a multi-part core extension for this use case:

https://four.htmx.org/extensions/hx-multipart

By @scriptogre

**Comment by Lanayx** [comment]
@1cg I don't think it's a replacement. Chunked encoding is just a regular html response, but without specifying content length, so server continues sending until the terminal chunk, it doesn't require multipart content type and special response encoding.
https://dev.to/sahan/deep-dive-how-chunked-transfer-encoding-works-4o9n

**Comment by carlos-verdes** [comment]
I agree with @Lanayx, actually this solution is the opposite of the problem I was trying to solve.
One example is to send candles for trading from the server, I don't want to send the full set of data in a single shot that's why I use chunked transfer, this helps both backend and frontend to avoid memory pressure on big data sets (as we just hold the next chunk in memory). 

Still this extension can be useful for some use cases and it's welcome.

**Comment by 1cg** [comment]
heya carlos, you can do exactly that with multi-part: stream down new content after an initial bit

i understand wanting to support chunked too and if someone wants to pick up and port the extension linked above (should be much cleaner in 4) I'd be happy to list it as a community extension for four, but multi-part, Web Socket & SSE are our core extensions for streaming HTML.
